lib/unison-sqlite/src/Unison/Sqlite/Connection.hs

```diff
@@ -0,0 +1,481 @@
+        Sqlite.bind statement params
+        callback (Sqlite.nextRow statement)
+
+------------------------------------------------------------------------------------------------------------------------
+-- Exceptions
+
+-- | A query was expected to return exactly one row, but it did not. The exception carries a string representation of
+-- the rows that were actually returned.
+newtype ExpectedExactlyOneRowException = ExpectedExactlyOneRowException
+  { rows :: String
+  }
+  deriving stock (Show)
+
+-- | A query was expected to return exactly one row, but it did not. The exception carries a string representation of
+-- the rows that were actually returned.
+newtype ExpectedAtMostOneRowException = ExpectedAtMostOneRowException
+  { rows :: String
+  }
+  deriving stock (Show)
+
+data SqliteExceptionInfo a = SqliteExceptionInfo
+  { sql :: Sql,
+    params :: Maybe a,
+    exception :: SomeShowTypeable,
+    connection :: Connection
+  }
+
+throwSqliteException :: SqliteExceptionInfo params -> IO a
+throwSqliteException SqliteExceptionInfo {connection, exception, params, sql} = do
+  threadId <- myThreadId
+  throwIO
+    SqliteException
+      { sql,
+        params = maybe "" anythingToString params,
+        exception,
+        connection,
+        threadId
+      }
+
+data SomeShowTypeable
+  = forall e. (Show e, Typeable e) => SomeShowTypeable e
+
+instance Show SomeShowTypeable where
+  show (SomeShowTypeable x) = show x
+
+-- | A @SqliteException@ represents an exception paired with some context that resulted in the exception.
+--
+-- A @SqliteException@ may result from a number of different conditions:
+--
+-- * The underlying sqlite library threw an exception, as when establishing a connection to a non-existent database.
+-- * A postcondition violation of a function like 'queryMaybe', which asserts that the resulting relation will have
+--   certain number of rows,
+-- * A postcondition violation of a function like 'queryListCheck', which takes a user-defined check as an argument.
+--
+-- A @SqliteException@ should not be inspected or used for control flow when run in a trusted environment, where the
+-- database can be assumed to be uncorrupt. Rather, wherever possible, the user of this library should write code that
+-- is guaranteed not to throw exceptions, by checking the necessary preconditions first. If that is not possible, it
+-- should be considered a bug in this library.
+--
+-- When actions are run on an untrusted codebase, e.g. one downloaded from a remote server, it is sufficient to catch
+-- just one exception type, @SqliteException@.
+data SqliteException = SqliteException
+  { sql :: Sql,
+    params :: String,
+    -- | The inner exception. It is intentionally not 'SomeException', so that calling code cannot accidentally
+    -- 'throwIO' domain-specific exception types, but must instead use a @*Check@ query variant.
+    exception :: SomeShowTypeable,
+    connection :: Connection,
```

Comment:
There's some useful information in the Connection object, but I don't know that we should be leaking the actual sqlite connection outside of its scope here.

E.g. this may cause a Sqlite Connection to propagate outside of a `withConnection` block can't it?

Reply:
Yeah, good point. How about `connection :: String` which `show`s the connection?

Reply:
I suppose it depends on why it's being included here, looks like the only info on the connection is `name` and `filepath`, so what if we just include `name` and `filepath` as fields here instead of the whole connection?

Reply:
Did you end up making any of these changes? Looks AFAICT it looks like the connection is still included.